geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/RouteSearchProcess.scala

```diff
@@ -0,0 +1,354 @@
+        throw new IllegalArgumentException(s"Route geometry field '$name' must be a LineString")
+      }
+      SelfClosingIterator(routes).map(_.getAttribute(index).asInstanceOf[LineString]).toSeq
+    }
+
+    val visitor = new RouteVisitor(routeGeoms, bufferSize, headingThreshold, bi, geomAttribute, isPoints, Option(headingField))
+
+    features.accepts(visitor, new NullProgressListener)
+    val results = visitor.getResult.results
+
+    configureOutput(features.getSchema, results, bins, binGeom, binDtg, binTrackId, binLabel)
+
+    results
+  }
+}
+
+class RouteVisitor(routes: Seq[LineString],
+                   routeBuffer: Double,
+                   threshold: Double,
+                   bidirectional: Boolean,
+                   geomAttribute: String,
+                   isPoints: Boolean,
+                   headingAttribute: Option[String]) extends FeatureCalc with FeatureAttributeVisitor with LazyLogging {
+
+  private var resultCalc: RouteResult = _
+
+  private val routeFilter =
+    orFilters(routes.map(ls => ff.dwithin(ff.property(geomAttribute), ff.literal(ls), routeBuffer, "meters")))
+
+  // for manual check, rewrite the filter to handle meters
+  // normally handled in our query planner, but we are going to use the filter directly here
+  private lazy val manualRouteFilter = routeFilter match {
+    case f: DWithin => new QueryPlanFilterVisitor(null).visit(f, null).asInstanceOf[Filter]
+    case f: Or      => new QueryPlanFilterVisitor(null).visit(f, null).asInstanceOf[Filter]
+  }
+
+  // for collecting results manually
+  private var manualCollection: ListFeatureCollection = _
+
+  override def getResult: RouteResult = resultCalc
+
+  // manually called for non-accumulo feature collections
+  override def visit(feature: Feature): Unit = {
+    val sf = feature.asInstanceOf[SimpleFeature]
+
+    if (manualCollection == null) {
+      manualCollection = new ListFeatureCollection(sf.getFeatureType)
+      resultCalc = RouteResult(matchRoutes(manualCollection))
+    }
+
+    if (manualRouteFilter.evaluate(sf)) {
+      manualCollection.add(sf)
+    }
+  }
+
+  // allows us to accept visitors from retyping feature collections
+  override def getExpressions: java.util.List[Expression] = {
+    import scala.collection.JavaConversions._
+    headingAttribute.map(ff.property).toSeq :+ ff.property(geomAttribute).asInstanceOf[Expression]
+  }
+
+  /**
+    * Optimized method to run distributed query. Sets the result, available from `getResult`
+    *
+    * @param source simple feature source
+    * @param query may contain additional filters to apply
+    */
+  def routeSearch(source: SimpleFeatureSource, query: Query): Unit = {
+    logger.debug(s"Visiting source type: ${source.getClass.getName}")
+    val result = if (routes.isEmpty) { source.getFeatures(Filter.EXCLUDE) } else {
+      import org.locationtech.geomesa.filter._
+      val filter = if (query == null || query.getFilter == Filter.INCLUDE) { routeFilter } else {
```

Comment:
As a corner case, do we need to consider any transforms or other query hints which might be set?

For instance, can I override the CBO in this process?  (And should I be able to?)